File: 78. Subsets/78. Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def backtrack(start, current):
            result.append(current[:])
            
            for i in range(start, len(nums)):
                current.append(nums[i])
                backtrack(i + 1, current)
                current.pop()

        result = []
        backtrack(0, [])
        return result

File: 78. Subsets/78. Subsets.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def subsets(nums)
    result = []
    backtrack(nums, 0, [], result)
    result
end

def backtrack(nums, start, current, result)
    result << current.dup
    
    (start...nums.length).each do |i|
        current << nums[i]
        backtrack(nums, i + 1, current, result)
        current.pop
    end
end